### ISSUE TITLE
OOM on creating threads to savePlayerData when migrating from flatfile to SQL database

### ISSUE THREAD

**Issue by RoboMWM** [GriefPrevention/GriefPrevention#20]
Reported by @Pixiio 

```
[01:08:02] [Server thread/INFO]: [GriefPrevention] Enabling GriefPrevention v15.3
[01:08:02] [Server thread/INFO]: [GriefPrevention] Finished loading configuration.
[01:08:03] [Server thread/INFO]: [GriefPrevention] Removing orphaned claim subdivision: Location{world=CraftWorld{name=world},x=9404.0,y=0.0,z=4705.0,pitch=0.0,yaw=0.0}
[01:08:03] [Server thread/INFO]: [GriefPrevention] Removing orphaned claim subdivision: Location{world=CraftWorld{name=world},x=3340.0,y=43.0,z=-2375.0,pitch=0.0,yaw=0.0}
[01:08:03] [Server thread/INFO]: [GriefPrevention] Removing orphaned claim subdivision: Location{world=CraftWorld{name=world},x=3339.0,y=43.0,z=-2378.0,pitch=0.0,yaw=0.0}
[01:08:03] [Server thread/INFO]: [GriefPrevention] Removing orphaned claim subdivision: Location{world=CraftWorld{name=world},x=7393.0,y=27.0,z=2203.0,pitch=0.0,yaw=0.0}
[01:08:03] [Server thread/INFO]: [GriefPrevention] 11663 total claims loaded.
[01:08:03] [Server thread/INFO]: [GriefPrevention] Customizable messages loaded.
[01:08:03] [Server thread/INFO]: [GriefPrevention] Successfully hooked into WorldGuard.
[01:08:03] [Server thread/INFO]: [GriefPrevention] There appears to be some data on the hard drive.  Migrating those data to the database...
[01:08:05] [Server thread/INFO]: [GriefPrevention] 11663 total claims loaded.
[01:08:05] [Server thread/INFO]: [GriefPrevention] Customizable messages loaded.
[01:08:05] [Server thread/INFO]: [GriefPrevention] Successfully hooked into WorldGuard.
[01:10:07] [Server thread/ERROR]: Error occurred while enabling GriefPrevention v15.3 (Is it up to date?)
java.lang.OutOfMemoryError: unable to create new native thread
    at java.lang.Thread.start0(Native Method) ~[?:1.8.0_72]
    at java.lang.Thread.start(Thread.java:714) ~[?:1.8.0_72]
    at me.ryanhamshire.GriefPrevention.DataStore.savePlayerData(DataStore.java:872) ~[?:?]
    at me.ryanhamshire.GriefPrevention.FlatFileDataStore.migrateData(FlatFileDataStore.java:821) ~[?:?]
    at me.ryanhamshire.GriefPrevention.GriefPrevention.onEnable(GriefPrevention.java:271) ~[?:?]
    at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:292) ~[patched_1.10.2.jar:git-Paper-878]
    at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:316) [patched_1.10.2.jar:git-Paper-878]
    at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:407) [patched_1.10.2.jar:git-Paper-878]
    at org.bukkit.craftbukkit.v1_10_R1.CraftServer.enablePlugin(CraftServer.java:375) [patched_1.10.2.jar:git-Paper-878]
    at org.bukkit.craftbukkit.v1_10_R1.CraftServer.enablePlugins(CraftServer.java:325) [patched_1.10.2.jar:git-Paper-878]
    at net.minecraft.server.v1_10_R1.MinecraftServer.t(MinecraftServer.java:417) [patched_1.10.2.jar:git-Paper-878]
    at net.minecraft.server.v1_10_R1.MinecraftServer.l(MinecraftServer.java:382) [patched_1.10.2.jar:git-Paper-878]
    at net.minecraft.server.v1_10_R1.MinecraftServer.a(MinecraftServer.java:328) [patched_1.10.2.jar:git-Paper-878]
    at net.minecraft.server.v1_10_R1.DedicatedServer.init(DedicatedServer.java:275) [patched_1.10.2.jar:git-Paper-878]
    at net.minecraft.server.v1_10_R1.MinecraftServer.run(MinecraftServer.java:583) [patched_1.10.2.jar:git-Paper-878]
    at java.lang.Thread.run(Thread.java:745) [?:1.8.0_72]
[01:10:07] [Server thread/INFO]: [GriefPrevention] Disabling GriefPrevention v15.3
[01:10:18] [Server thread/INFO]: [GriefPrevention] GriefPrevention disabled.
```

**Comment by RoboMWM** [comment]
Please test this debug build. Rename the .zip extension to .jar

[GriefPrevention-debug.zip](https://github.com/TechFortress/GriefPrevention/files/515263/GriefPrevention-debug.zip)

Paste the server log here after running it.

**Comment by Pixiio** [comment]
Hey. Thanks for helping. The reason why we are converting is because sometimes players just get claimblocks out of no where. We've been running Griefprevention since a very long time (years) on this server, so it might be because there have been some changes in file setup etc. On our other server with GP (that is newer) there are no problems at all.

Here is our config: http://pastebin.com/aDhZq5X3
Random player file: http://i.imgur.com/AMXdCT0.png
Random claim file: http://i.imgur.com/SfXtS7w.png

The converting was tested on a dedicated server. Also tryed it on my own computer, same problem.

Thanks in advance.

**Comment by RoboMWM** [comment]
> The reason why we are converting is because sometimes players just get claimblocks out of no where. 

Please elaborate on this; what do you mean exactly by this? Can you give a scenario?

Also enable debug mode in the config - and check the GP logs whenever you notice a player "getting claimblocks out of nowhere."

**Comment by Pixiio** [comment]
We're having reports of players getting claimblocks out of nowhere. These reports have been here for a long time, but they are not that ofen. A player on the server got 200.000 claim blocks last week. When I searched in the ''normal'' logs for ''adjusted claimblocks'' I couldn't find anything weird.

**Comment by RoboMWM** [comment]
This isn't enough information. Enable debug logging so you can see claim blocks being delivered. Searching for "adjusted claimblocks" isn't sufficient.

**Comment by Pixiio** [comment]
I'll take a look, thanks!

**Comment by Pixiio** [comment]
Problem has been fixed, it was a problem on our end. A long time ago we did some test with the ACB permission but it looks like we've never removed it... The reason why I didn't though about this because people said they were getting random amounts of claimblocks.. 

About the MYSQL migration: don't need it anymore. 

Thanks alot for the help!

**Comment by RoboMWM** [comment]
Thanks for letting me know! I'm going to be reworking the "extra claim block" accrual to be an API feature, and remove it from the main plugin, since as it stands now it's hardcoded anyways.

**Comment by Pixiio** [comment]
Still getting reports of players getting random claimblocks.
I'll enable debug again.

Am I the only one with this problem? It looks like we have it on both our survival servers with GP.

**Comment by Pixiio** [comment]
So I found the problem. The players that have random claimblocks got random accruedblocks.
Don't really know how that is possible because:

    Claim Blocks Accrued Per Hour:
      Default: 0
      With 'fasteraccrual' Permission: 110
      With 'fastestaccrual' Permission: 125
    Max Accrued Claim Blocks:
      Default: 250
      With 'moreaccrued' Permission: 5000
      With 'mostaccrued' Permission: 10000
    AbandonReturnRatio: 1.0

**Comment by Pixiio** [comment]
@RoboMWM This bug (the migration bug) is still here. Tryed it on both our Survival servers. It only migrates some data to the database and then gives this bug. (we migrated 1 server now but with alot of data loss)

Debug didn't show much.

**Comment by RoboMWM** [comment]
Did the server log show anything with the attached debug build?

**Comment by RoboMWM** [comment]
http://stackoverflow.com/questions/5253072/java-unable-to-create-new-native-thread

Try some of the things suggested here (increasing max files that can be opened, decreasing JVM size, etc.)

**Comment by Pixiio** [comment]
I've tryed more things, but i'm pretty sure it's not the server. I've tryed it on 2 different dedicated servers and even my own computer (with a local mysql).

**Comment by RoboMWM** [comment]
> more things

What things.

Well, definitely still need more information. Like number of claims and playerdata files, etc.

And you're doing this with _only_ GP installed, yes?

**Comment by Pixiio** [comment]
I do have other plugins installed. I've got a nice debug log that you can probably use. :)

http://pastebin.com/sq5KYPr2

**Comment by RoboMWM** [comment]
Try doing the migration without other plugins installed...

**Comment by Pixiio** [comment]
Same issue. Is it normal that so many threads are starting? O.o

**Comment by RoboMWM** [comment]
> Well, definitely still need more information. Like number of claims and playerdata files, etc.

**Comment by Pixiio** [comment]
Number of claims: 11.368 claimfiles
Number of playerfiles: 133.593 playerfiles

**Comment by RoboMWM** [comment]
So, I haven't worked with SQL or threads all that much, so brainstorming on ideas out loud:

- Schedule a max of x threads per tick, or per 5 second intervals(?)
- Schedule more than 1 save per thread?
- Make an offline converter?

**Comment by Pixiio** [comment]
I'm not into SQL or threads either so maybe someone else has an idea..

**Comment by Ceees2** [comment]
Why not use an SQL pool and create a queue. For example like LogBlock:
https://github.com/LogBlock/LogBlock/blob/master/src/main/java/de/diddiz/util/MySQLConnectionPool.java
https://github.com/LogBlock/LogBlock/blob/master/src/main/java/de/diddiz/LogBlock/Consumer.java

**Comment by RoboMWM** [comment]
Feel free to PR. I'm not all that familiar with SQL yet, but I'll spend time on it when I get said time - which isn't right now unfortunately :/

I'd like to avoid having to implement an entire pool/consumer sort of thing for just a simple conversion - at that point, I might as well make some sort of offline tool.

**Comment by RoboMWM** [comment]
Given how there's no dedicated Converter class in this project or the like, it's gonna be an ugly mess to change the design of this converter. What it does is load all the playerdata files, then save each one (to the MySQL storage) asynchronously. In doing so it's going to generate a bunch of threads. Would be better to queue these up and save all at once in a thread instead - but this type of save would have to be implemented, since the current one only saves one at a time, one per thread.

If anyone wants to poke at this they can, but given how GP works and what data it needs, using SQL as its backend doesn't make a lot of sense in the first place.

**Comment by RoboMWM** [comment]
As for the reason to convert, using SQL (as seen in other issues) won't fix the "random claimblocks" issue, which was recently fixed in 16.9 (caused by abandonReturnRatio being set to anything other than 1 and administrators abandoning others' claims.)
